pkg/tempo/internal/tempoutils/server.go

```diff
@@ -73,121 +73,144 @@ func NewServerWithRandomPort(callback func(pdata.Traces)) (srv *Server, addr str
-	exportersFactory, err := component.MakeExporterFactoryMap()
+	exportersFactory, err := component.MakeExporterFactoryMap(otlpexporter.NewFactory())
 	if err != nil {
-		return nil, fmt.Errorf("failed to make exporter factory map: %w", err)
+		return component.Factories{}, fmt.Errorf("failed to make exporter factory map: %w", err)
 	}
 
 	processorsFactory, err := component.MakeProcessorFactoryMap(
 		newFuncProcessorFactory(callback),
 	)
 	if err != nil {
-		return nil, fmt.Errorf("failed to make processor factory map: %w", err)
+		return component.Factories{}, fmt.Errorf("failed to make processor factory map: %w", err)
 	}
 
-	factories := component.Factories{
+	return component.Factories{
 		Extensions: extensionsFactory,
 		Receivers:  receiversFactory,
 		Processors: processorsFactory,
 		Exporters:  exportersFactory,
-	}
-	otelCfg, err := config.Load(v, factories)
-	if err != nil {
-		return nil, fmt.Errorf("failed to make otel config: %w", err)
-	}
+	}, nil
+}
 
-	var (
-		logger    = zap.NewNop()
-		startInfo component.ApplicationStartInfo
-	)
+func configFactory(addr string) func(v *viper.Viper, cmd *cobra.Command, factories component.Factories) (*configmodels.Config, error) {
+	return func(v *viper.Viper, cmd *cobra.Command, factories component.Factories) (*configmodels.Config, error) {
+		conf := util.Untab(fmt.Sprintf(`
+processors:
+	func_processor:
+receivers:
+  otlp:
+		protocols:
+			grpc:
+				endpoint: %s
+exporters:
+  otlp:
+    endpoint: example.com:12345
+service:
+	pipelines:
+		traces:
+			receivers: [otlp]
+			processors: [func_processor]
+			exporters: [otlp]
+	`, addr))
 
-	exporters, err := builder.NewExportersBuilder(logger, startInfo, otelCfg, factories.Exporters).Build()
-	if err != nil {
-		return nil, fmt.Errorf("failed to build exporters: %w", err)
-	}
-	if err := exporters.StartAll(context.Background(), nil); err != nil {
-		return nil, fmt.Errorf("failed to start exporters: %w", err)
-	}
+		var cfg map[string]interface{}
+		if err := yaml.NewDecoder(strings.NewReader(conf)).Decode(&cfg); err != nil {
+			panic("could not decode config: " + err.Error())
+		}
 
-	pipelines, err := builder.NewPipelinesBuilder(logger, startInfo, otelCfg, exporters, factories.Processors).Build()
-	if err != nil {
-		return nil, fmt.Errorf("failed to build pipelines: %w", err)
-	}
-	if err := pipelines.StartProcessors(context.Background(), nil); err != nil {
-		return nil, fmt.Errorf("failed to start pipelines: %w", err)
-	}
+		if err := v.MergeConfigMap(cfg); err != nil {
+			return nil, fmt.Errorf("failed to merge in mapstructure config: %w", err)
+		}
 
-	receivers, err := builder.NewReceiversBuilder(logger, startInfo, otelCfg, pipelines, factories.Receivers).Build()
-	if err != nil {
-		return nil, fmt.Errorf("failed to build receivers: %w", err)
-	}
-	if err := receivers.StartAll(context.Background(), nil); err != nil {
-		return nil, fmt.Errorf("failed to start receivers: %w", err)
+		otelCfg, err := config.Load(v, factories)
```

Comment:
The docs ask to verify the config using `ValidateConfig`. I'm not sure if the config is validated later but might be helpful to fail fast here?
```go
// Load loads a Config from Viper.
// After loading the config, need to check if it is valid by calling `ValidateConfig`.
func Load(
  ...
```

Reply:
Thanks for pointing it out. The interface has changed in v0.27, so it no longer applies.